### PR TITLE
ICU-21248 Adds source file check (UTF-8 and absence of BOM) to

### PR DESCRIPTION
**Pull request by gnrunge**
Travis Continued Integration.

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21248
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [ ] Tests included
- [ ] Documentation is changed or added

Screen Shot 2020-09-17 at 3 33 31 PM

### CODE REVIEW

**Comment by gnrunge** [comment]
Question: ICU also has the icu/tools/scripts/ directory, which may be a better location for this script. Any comments?

File: icu4c/source/tools/icu-file-utf8-check.py [hunk 1]
```diff
@@ -22,6 +22,8 @@
 #  The tool operates recursively on the directory from which it is run.
 #  Only files from the ICU github repository are checked.
 #  No changes are made to the repository; only the working copy will be altered.
+#  The script checks all source files and keeps a non-zero exit code if any of
```
Comment (jefgen) [inline icu4c/source/tools/icu-file-utf8-check.py:25]: nit: For the phrase, "and keeps a non-zero exit code", would "and returns a non-zero exit code" maybe be a slightly more common way to say it? I'm not sure...
(I see the script does have a rc variable, so it does "keep" track of the return code, but a caller of the program might not care I guess.)

File: icu4c/source/tools/icu-file-utf8-check.py [hunk 2]
```diff
@@ -61,16 +67,20 @@ def check_file(file_name, is_source):
             bytes.decode("UTF-8")
         except UnicodeDecodeError:
             print("Error: %s is a source code file but contains non-utf-8 bytes." % file_name)
-    
+            rc = 1
+
     if bytes[0] == 0xef:
         if not (file_name.endswith(".txt") or file_name.endswith(".sln")
-                    or file_name.endswith(".targets")
+                    or file_name.endswith(".targets") or file_name.endswith(".md")
```
Comment (jefgen) [inline icu4c/source/tools/icu-file-utf8-check.py:74]: Are the markdown files all UTF-8 with BOM? I wonder if we might want to remove the BOM on them too...

File: icu4c/source/tools/icu-file-utf8-check.py [hunk 3]
```diff
@@ -84,23 +94,30 @@ def main(argv):
...
-        check_file(f, source_file)
+        sys.exit(2)
+
+    git_cmd = "git ls-files DIR"
+    for dir in icu_directories_to_be_scanned:
+        print('Scanning ' + dir)
+        output = runCommand(git_cmd.replace("DIR", dir))
+        file_list = output.splitlines()
+
+        source_file_re = re.compile(".*((?:\\.c$)|(?:\\.cpp$)|(?:\\.h$)|(?:\\.java$))")
```
Comment (jefgen) [inline icu4c/source/tools/icu-file-utf8-check.py:105]: I wonder if this regex could be moved outside of the for loop.
(Not that perf is really that big of a concern here though...)

**Comment by jefgen** [comment]
> Question: ICU also has the icu/tools/scripts/ directory, which may be a better location for this script. Any comments?

Good question. 
FWIW, I think having this under `icu4c/source` seems a bit odd since it's used for checking `icu4j` as well.

The Git/Jira commit checker script lives under `icu/tools/commit-checker/`.
I wonder if perhaps something like `icu/tools/utf8-checker/` ?

**Comment by markusicu** [comment]
> > Question: ICU also has the icu/tools/scripts/ directory, which may be a better location for this script. Any comments?
> 
> Good question.
> FWIW, I think having this under `icu4c/source` seems a bit odd since it's used for checking `icu4j` as well.
> 
> The Git/Jira commit checker script lives under `icu/tools/commit-checker/`.
> I wonder if perhaps something like `icu/tools/utf8-checker/` ?

https://github.com/unicode-org/icu/tree/master/tools/scripts is definitely a better place.
As a single script I don't think it needs a new subfolder.

File: icu4c/source/tools/icu-file-utf8-check.py (hunk 2 above)
Comment (markusicu) [reply icu4c/source/tools/icu-file-utf8-check.py:74]: Hopefully the Markdown files do *not* start with a BOM.

Comment (jefgen) [reply icu4c/source/tools/icu-file-utf8-check.py:74]: Right -- I don't know if we want to start skipping the `.md` files.

Comment (gnrunge) [reply icu4c/source/tools/icu-file-utf8-check.py:74]: The script flagged one MD file:
icu4c/source/test/testdata/break_rules/README.md
I thought it's close enough to a .txt file and added .md to the filter. Apparently MD files are not that close after all.

File: icu4c/source/tools/icu-file-utf8-check.py (hunk 1 above)
Comment (gnrunge) [reply icu4c/source/tools/icu-file-utf8-check.py]: I like your wording - done.

File: icu4c/source/tools/icu-file-utf8-check.py (hunk 3 above)
Comment (gnrunge) [reply icu4c/source/tools/icu-file-utf8-check.py]: Valid point -done.

**Comment by gnrunge** [comment]
Please take a look again. The BOM is gone from the README.md, albeit not visible in the diff view.
Thx.

**Review by srl295 — DISMISSED** [review]

File: tools/scripts/icu-file-utf8-check.py [hunk 4]
```diff
@@ -32,6 +34,9 @@
 import getopt
 
 
+# List of directories to check for UTF-8 and BOM. Extend as needed.
+icu_directories_to_be_scanned = ["icu4c/source", "icu4j"]
```
Comment (markusicu) [inline tools/scripts/icu-file-utf8-check.py:38]: Does it work if we scan all files? Or at least "tools"? Or is there too much amiss right now?

**Review by markusicu — DISMISSED** [review]
lgtm

**Comment by gnrunge** [comment]
No findings reported when executed from the root icu/ directory on all the
9,267 files listed by git ls-files.

**Comment by markusicu** [comment]
> No findings reported when executed from the root icu/ directory on all the 9,267 files listed by git ls-files.

Well, maybe change the list of directories to just `["."]`? Keep it as a list, so we can go back to specific folders if needed.

**Comment by gnrunge** [comment]
Coverage of UTF-8/BOM check now all of icu/.

**Review by markusicu — APPROVED** [review]
lgtm tnx!

**Review by jefgen — APPROVED** [review]
Thanks @gnrunge! :)

**Pull request merged**